app/src/main/java/ch/epfl/balelecbud/EmergencyNumbersActivity.java

```diff
@@ -0,0 +1,103 @@
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+import ch.epfl.balelecbud.Emergency.EmergencyInfo;
+import ch.epfl.balelecbud.Emergency.EmergencyNumbers;
+
+public class EmergencyNumbersActivity extends BasicActivity {
+
+    private ListView numbersListView;
+    private ArrayAdapter arrayAdapter;
+    private Map<String, String> repertoryMap = new HashMap<String, String>();
+
+    private FirebaseFirestore db = FirebaseFirestore.getInstance();
+    private CollectionReference emergencyNumbersRef = db.collection("emergencyNumbers");
+
+
+    @Override
+    protected void onCreate(Bundle savedInstanceState) {
+        super.onCreate(savedInstanceState);
+        setContentView(R.layout.activity_emergency_numbers);
+        numbersListView = findViewById(R.id.numbersListView);
+        arrayAdapter = new ArrayAdapter(
+                this,
+                android.R.layout.simple_list_item_1,
+                new ArrayList<String>());
+    }
+
+    @Override
+    protected void onStart() {
+        super.onStart();
+        emergencyNumbersRef.addSnapshotListener(this, new EventListener<QuerySnapshot>() {
+            @Override
+            public void onEvent(QuerySnapshot queryDocumentSnapshots, FirebaseFirestoreException e) {
+                if (e != null) {
+                    return;
+                }
+                for (QueryDocumentSnapshot documentSnapshot : queryDocumentSnapshots) {
+                    EmergencyNumbers number = documentSnapshot.toObject(EmergencyNumbers.class);
+                    repertoryMap.put(number.getName(), number.getNumber());
+                }
+
+                List<String> numberList = new ArrayList<String>(Arrays.asList(repertoryMap.keySet().toArray(new String[0])));
+                arrayAdapter = new ArrayAdapter(EmergencyNumbersActivity.this, android.R.layout.simple_list_item_1, numberList);
+                numbersListView.setAdapter(arrayAdapter);
+
+                numbersListView.setOnItemClickListener(new AdapterView.OnItemClickListener() {
+                    @SuppressLint("MissingPermission")
+                    @Override
+                    public void onItemClick(AdapterView<?> parent, View view, int position,
+                                            long id) {
+                        String entry = (String) parent.getAdapter().getItem(position);
+                        Intent intent = new Intent(Intent.ACTION_CALL, Uri.parse("tel:" + repertoryMap.get(entry)));
+
+                        String permissions[] = {Manifest.permission.CALL_PHONE};
+                        int PERMISSION_TO_CALL = 0;
+
+                        while (ActivityCompat.checkSelfPermission(EmergencyNumbersActivity.this,
+                                Manifest.permission.CALL_PHONE) != PackageManager.PERMISSION_GRANTED) {
+                            ActivityCompat.requestPermissions(EmergencyNumbersActivity.this,
```

Comment:
Doesn't it prevent the user to cancel the call ? And busy waiting is not a good solution maybe consider using [onRequestPermissionsResult()](https://developer.android.com/reference/androidx/core/app/ActivityCompat.OnRequestPermissionsResultCallback#onRequestPermissionsResult(int,%20java.lang.String[],%20int[])), you can read more about it [here](https://developer.android.com/training/permissions/requesting#handle-response)

Reply:
Fixed in https://github.com/karimassi/balelec-bud/pull/31/commits/013bf3802330f510f2fa1b1f8e896a505cfceef2